rlpx/rlpx.go

```diff
@@ -0,0 +1,260 @@
+// Implements the RLPx protocol. Code is based off the devp2p spec defined here:
+// https://github.com/ethereum/devp2p/blob/master/rlpx.md
+package rlpx
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/hmac"
+	"errors"
+	"fmt"
+	"hash"
+	"net"
+
+	"github.com/decred/dcrd/dcrec/secp256k1/v4"
+	"golang.org/x/crypto/sha3"
+
+	"github.com/indexsupply/x/bint"
+	"github.com/indexsupply/x/enr"
+	"github.com/indexsupply/x/isxerrors"
+	"github.com/indexsupply/x/isxhash"
+	"github.com/indexsupply/x/rlp"
+)
+
+type mstate struct {
+	block  cipher.Block
+	hash   hash.Hash
+	stream cipher.Stream
+}
+
+func newmstate(ke, km []byte) (*mstate, error) {
+	eb, err := aes.NewCipher(ke)
+	if err != nil {
+		return nil, err
+	}
+	es := cipher.NewCTR(eb, make([]byte, 16))
+
+	mb, err := aes.NewCipher(km)
+	if err != nil {
+		return nil, err
+	}
+	h := sha3.NewLegacyKeccak256()
+	return &mstate{
+		hash:   h,
+		block:  mb,
+		stream: es,
+	}, nil
+}
+
+// updates s.emac (egress mac) using the following devp2p construction:
+//
+// header-mac-seed = aes(mac-secret, keccak256.digest(egress-mac)[:16]) ^ header-ciphertext
+// egress-mac = keccak256.update(egress-mac, header-mac-seed)
+// header-mac = keccak256.digest(egress-mac)[:16]
+func (ms *mstate) header(h []byte) []byte {
+	prev := ms.hash.Sum(nil)
+	dest := make([]byte, 16)
+	ms.block.Encrypt(dest, prev[:16])
+	for i := range dest {
+		dest[i] ^= h[i]
+	}
+	ms.hash.Write(dest)
+	return ms.hash.Sum(nil)[:16]
+}
+
+// updates s.emac (egress mac) using the following devp2p construction:
+//
+// egress-mac = keccak256.update(egress-mac, frame-ciphertext)
+// frame-mac-seed = aes(mac-secret, keccak256.digest(egress-mac)[:16]) ^ keccak256.digest(egress-mac)[:16]
+// egress-mac = keccak256.update(egress-mac, frame-mac-seed)
+// frame-mac = keccak256.digest(egress-mac)[:16]
+func (ms *mstate) frame(fct []byte) []byte {
+	ms.hash.Write(fct)
+	prev := ms.hash.Sum(nil)
+	dest := make([]byte, 16)
+	ms.block.Encrypt(dest, prev[:16])
+	for i := range dest {
+		dest[i] ^= prev[i]
+	}
+	ms.hash.Write(dest)
+	return ms.hash.Sum(nil)[:16]
+}
+
+type session struct {
+	Verbose bool
+
+	conn          *net.TCPConn
+	local, remote *enr.Record
+	ig, eg        *mstate
+}
+
+func (s *session) log(format string, args ...any) {
+	if s.Verbose {
+		fmt.Printf(format, args...)
+	}
+}
+
+type handshake struct {
```

Comment:
@lechengfan this can be our integration point

Reply:
Yeah, that sounds great! I think at this point, my PR is able to fill in `initNonce`, `respNonce`, `localEphPrvKey` and `remoteEphPubKey` so that's sufficient for you to generate the shared secrets.